### ISSUE TITLE
DPatch generate function don't update its patched image on the iteration loop

### ISSUE THREAD

**Issue by naufalso** [Trusted-AI/adversarial-robustness-toolbox#1047]
**Describe the bug**
I tried to run the DPatch attack on my custom TensorFlow object detection estimator and I found that gradient loss from the estimator used to update the DPatch is kept constant / not changing.

Then I found that the DPatch generate function on the iteration loop doesn't update its patched image.

Please check the following snippet code of art/attacks/evasion/dpatch.py on line 128 to 222
  ```
         for i_step in trange(self.max_iter, desc="DPatch iteration", disable=not self.verbose):
            if i_step == 0 or (i_step + 1) % 100 == 0:
                logger.info("Training Step: %i", i_step + 1)

            num_batches = math.ceil(x.shape[0] / self.batch_size)
            patch_gradients = np.zeros_like(self._patch)

            for i_batch in range(num_batches):
                i_batch_start = i_batch * self.batch_size
                i_batch_end = min((i_batch + 1) * self.batch_size, patched_images.shape[0])

                gradients = self.estimator.loss_gradient(
                    x=patched_images[i_batch_start:i_batch_end], y=patch_target[i_batch_start:i_batch_end],
                ) 

                for i_image in range(gradients.shape[0]):

                    i_x_1 = transforms[i_batch_start + i_image]["i_x_1"]
                    i_x_2 = transforms[i_batch_start + i_image]["i_x_2"]
                    i_y_1 = transforms[i_batch_start + i_image]["i_y_1"]
                    i_y_2 = transforms[i_batch_start + i_image]["i_y_2"]

                    if self.estimator.channels_first:
                        patch_gradients_i = gradients[i_image, :, i_x_1:i_x_2, i_y_1:i_y_2]
                    else:
                        patch_gradients_i = gradients[i_image, i_x_1:i_x_2, i_y_1:i_y_2, :]

                    patch_gradients = patch_gradients + patch_gradients_i

            if self.target_label:
                self._patch = self._patch - np.sign(patch_gradients) * self.learning_rate
            else:
                self._patch = self._patch + np.sign(patch_gradients) * self.learning_rate

            if self.estimator.clip_values is not None:
                self._patch = np.clip(
                    self._patch, a_min=self.estimator.clip_values[0], a_max=self.estimator.clip_values[1],
                )
```

The **patched_images** variable is not updated on the iteration loop, since it used for updating the gradients on lines 196 - 198.

**Expected behavior**
**patched_images** should be updated before calling self.estimator.loss_gradient

**System information (please complete the following information):**
 - OS: Ubuntu Docker
 - Python version: Python 3.6.9
 - ART version or commit number: main branch, commit: 8981dd0ff55b7c9303355aaad4087e1f2a2f2daa
 - TensorFlow / Keras / PyTorch / MXNet version: Tensorflow 2.5.0-dev20210204

**Comment by beat-buesser** [comment]
Hi @naufalso Thank you very much for using ART and raising this issue! I think you are right and we should add an update for `patched_images` inside of the loop over the optimisation iterations.

Would you be interested to investigate a possible solution leading to a pull request? I would be happy to provide support.

**Comment by beat-buesser** [comment]
Hi @naufalso I have pushed a possible solution to branch `development_issue_1047` in #1049 . If you have time, could you please take a look and let me know what you think?

**Comment by naufalso** [comment]
Hi @beat-buesser, Thanks for the fast response. 
I have checked on your PR codes and they looks great.
I previously forget to mention about the possible optional mask initialization bugs and it looks like you also has fixed it (I previously just create a dummy Mask to avoid the initialization error then I found another critical bugs that patched images is not updated).

I will try to run your updated code for the verification and recheck that everything is running as expected. 

Thanks a lot for your fast response.

**Comment by beat-buesser** [comment]
Hi @naufalso Great, thank you! Yes, please let me know what your experiments find out and please keep finding and reporting bug!

Have you also noticed the attack `RobustDPatch` in ART? It comes with several improvements over `DPatch`.

**Comment by naufalso** [comment]
Hi @beat-buesser, Thanks.
Yes, I have tried the `RobustDPatch` after I found the bug on DPatch. And It works pretty well 👍.

I just retested your updated code and it works as expected (but of course it is not as good as RobustDPatch) with the default untargeted-attack setting (when `target-label` parameter on generate function is not defined). 
However, when I set the target label I found other possible bugs.

## Bugs when target-label parameter is set

I am wondering what is the Ground Truth Box format of the ART standard?

On your updated DPatch code line 176
`target_dict["boxes"] = np.asarray([[i_x_1, i_y_1, i_x_2, i_y_2]])`

It shows that the ground truth format is `x1, y1, x2, y2` or `x_min, y_min, x_max, y_max`, is that right?
But when I plot the `patched_images[0]` and print the `patch_target[0]['boxes']` that is called in code line 213~
```
gradients = self.estimator.loss_gradient(
      x=patched_images[i_batch_start:i_batch_end],
      y=patch_target[i_batch_start:i_batch_end],
  )
```
I found that the ground truth boxes is in `y1, x1, y2, x2` or `y_min, x_min, y_max, x_max`.
Please check this screenshot:
![image](https://user-images.githubusercontent.com/51362638/114801234-48826780-9dd6-11eb-8c00-706e486a0fb9.png)

So which format is the correct format?
- `x1, y1, x2, y2` or  `y1, x1, y2, x2` ?
- Does the image assumed to have (H,W) or (W,H) format ?

If I checked the code, I think the code has written in the wrong variable names (between x and y) if it uses standard (H, W) image format.
I think the root problem is from the code lines 298-298:
```
i_x_1 = random.randint(0, x_copy.shape[1] - 1 - patch_copy.shape[0])
i_y_1 = random.randint(0, x_copy.shape[2] - 1 - patch_copy.shape[1])
```
Because if the image format is standard (H, W) `x_copy.shape[1]` and `patch_copy.shape[0]` is referred to the height (y),
while `x_copy.shape[2]` and `patch_copy.shape[1]` is referred to width (x)

The variable naming problem can also be seen on code line 225-228:
```
if self.estimator.channels_first:
    patch_gradients_i = gradients[i_image, :, i_x_1:i_x_2, i_y_1:i_y_2]
else:
    patch_gradients_i = gradients[i_image, i_x_1:i_x_2, i_y_1:i_y_2, :]
```
If we use (H, W) standard image format, the cropping should be [i_y_1:i_y_2, i_x_1:i_x2]
Or does ART really assumed that all images are using (W, H) format, not normal NumPy (H, W) format?

Note: all code line is based on updated PR codes

## Other x, y naming confusion:

I found this x, y naming confusion when I am trying to implement the custom object detection estimator for TF2.

The loss_gradient function on art/estimators/object_detection/pytorch_faster_rcnn.py (line 157) told us that ground truth box format is:
`- boxes (FloatTensor[N, 4]): the predicted boxes in [x1, y1, x2, y2] format, with values between 0 and H and 0 and W`

while the loss_gradient function on art/estimators/object_detection/tensorflow_faster_rcnn.py (line 326~) told us that the ground truth box format is:
`- 'boxes': A list of 'nb_samples' size of 2-D tf.float32 tensors of shape [num_boxes, 4] containing
  coordinates of the groundtruth boxes. Groundtruth boxes are provided in [y_min, x_min, y_max, x_max]
  format and also assumed to be normalized as well as clipped relative to the image window with
  conditions y_min <= y_max and x_min <= x_max.`

Please help me clarify this naming confusion, thanks a lot.

**Comment by beat-buesser** [comment]
Hi @naufalso

ART tries to follow the following pattern:
- the images are either in format `NHWC` or `NCHW` with `H` always on a smaller index than `W`
- the boxes of object detector predictions of ART estimators should be `[x0, y0, x1, y1]` where `x` is in range [0, `H`] and `y` in range [0, `W`]. This leads to x-coordinates being in vertical and y-coordinates being in horizontal direction on the image, which is contrary to the more common use of x-coordinates in horizontal and y-coordinates in vertical direction.
- the docstring of `art.estimators.object_detection.tensorflow_faster_rcnn` should be changed from ` [y_min, x_min, y_max, x_max]` to `[x_min, y_min, x_max, y_max]`.

Are you training your own object detection model?

**Comment by naufalso** [comment]
Hi @beat-buesser, ah then it is not a bug. The problem comes from my misunderstanding.
Thanks a lot for the clear explanation. 

If that is the case, then I think we can close this issue since you have fixed it in your PR :)

Yes, I create my own object detection models based on TF Object Detection API v2 and Tensorflow v2 Keras for YOLOv4.
Since the current ART only supports TF Object Detection API v1 and PyTorch for FasterRCNN, I develop a new estimator based on those two codes and overriding the required methods.

**Comment by beat-buesser** [comment]
Hi @naufalso 

Your project creating estimators for new object detection models like TF Object Detection API v2 and Tensorflow v2 Keras for YOLOv4 sounds very interesting. Would you be interested to contribute them to ART at some point when your project allows it? If yes, we could open an new feature issue to plan for the integration of your new tools for you and we would be happy to provide further support.

We have followed with ART's definition of x/y/H/W the definitions in the docstrings of https://pytorch.org/vision/stable/_modules/torchvision/models/detection/faster_rcnn.html, but we have noticed that there might be a confusion there: instead of `0 <= x1 < x2 <= W` we observed `0 <= x1 < x2 <= H`. If you have a chance to compare with `torch`'s FasterRCNN I would be interested to know what you find out.

**Comment by naufalso** [comment]
Hi @beat-buesser, sorry for the late reply.

Unfortunately, my project is from an institution which security-sensitive. But as personal, I can contribute to writing that modules which follow ART standard. I am interested but I just can't guarantee when I can finish it.

Sorry, I am not so familiar with PyTorch because I always use TensorFlow a lot. But from your shared docstring, I am still assuming that the x-axis is a horizontal axis that has maximum Width, while the y-axis is a vertical axis that has maximum Height. 

[x1, y1, x2, y2] or [top_left, bottom_right] is the common approach for defining the bounding box. While the common other approach is [x, y, w, h].

My suggestion, if the ART using a different standard than the common approach, I think the documentation should be clear and not ambiguous.

**Comment by beat-buesser** [comment]
Hi @naufalso

I agree with your observation and have pushed updates to #1069 to be released with ART 1.6.2. These updates aim to ensure that all attacks are treating object detection labels internally as of PyTorch format ([x1, y1, x2, y2] format, with 0 <= x1 < x2 <= W and 0 <= y1 < y2 <= H), but can accept labels in TensorFlow format in method their `generate` ([y1, x1, y2, x2] format, with 0 <= x1 < x2 <= W=1.0 and 0 <= y1 < y2 <= H=0.1) if the estimator's model is expecting TensorFlow format, defined by their property `native_label_is_pytorch_format`.

The object detection estimators in `art.estimators.object_detection` now all support the argument `standardise_output` in methods `predict` and `loss_gradients` to enable them accepting and returning labels in PyTorch format, respectively, even if their model expects TensorFlow format.

If you have time to take a look at #1069, please let me know what you think!
